espnet/lm/pytorch_backend/lm.py

```diff
@@ -307,14 +95,13 @@ def update_core(self):
         # self.converter does this job
         # (it is chainer.dataset.concat_examples by default)
         x, t = concat_examples(batch, device=self.device, padding=(0, -100))
-        loss, count = self.model(x, t)
-        loss = loss.sum()
-        count = count.sum()
-        reporter.report({'loss': float(loss.detach() / count)}, optimizer.target)
-        reporter.report({'count': int(count)}, optimizer.target)
+        loss, logp, count = self.model(x, t)
+        reporter.report({'train_loss': float(loss.mean())}, optimizer.target)
+        reporter.report({'loss': float(logp.sum())}, optimizer.target)
```

Comment:
Strongly agree. Let me explain why I made these three return values. It is complicated.

1. `loss`: this is only used for `loss.mean().backward()` to train NN. It can be different from logp because of normalization.
2. `logp`: this is used as numerator of perplexity. I rethink its should be named to `nll` (negative log likelihood) because it is negative.
3. `count`: this is used as the denominator of perplexity.

To rename the report key 'loss' into `nll` needs to change `compute_perplexity` function and  chainer backend as well. However, I will try that.

The reason why evaluator puts `logp / count` in `'loss'`
```python
reporter.report({'loss': logp / count}, self.model.reporter)
```
is that evaluator does not use `compute_perplexity` that computes ppl with averaged `results['main/loss'] / results['main/count']` online. So I think the code above should be something like:
```python
reporter.report({'test_perplexity': logp / count}, self.model.reporter)
```

Reply:
I have fixed this.